### ISSUE TITLE
extra spaces in latex output from tables with cell occupying two or more rows

### ISSUE THREAD

**Issue by jfbu** [sphinx-doc/sphinx#2723]
Example:

``` rst
Welcome to Test's documentation!
================================

Contents:

.. toctree::
   :maxdepth: 2

+--------+-------------------+
| N      | Location Location |
|        +-------------------+
|        | IP Address        |
+========+===================+
| foobar | there             |
|        +-------------------+
|        | xxx.yyy.zzz.ttt   |
+--------+-------------------+
```

output

![screenshot1](https://cloud.githubusercontent.com/assets/2589111/16451643/9486a88a-3e04-11e6-8643-613ff694ac7f.png)

should rather be something like

![screenshot2](https://cloud.githubusercontent.com/assets/2589111/16451655/a7e2aeba-3e04-11e6-8ae9-53ac83823872.png)

Caused by this kind of mark-up in produced latex file:

``` latex
\hline \multirow{2}{*}{
foobar
} & 
```

the complete table is

``` latex
\noindent\begin{tabulary}{\linewidth}{|L|L|}
\hline
 \multirow{2}{*}{\textsf{\relax 
N
}} & \textsf{\relax 
Location Location
}\\
\cline{2-2} & \textsf{\relax 
IP Address
}\\
\hline \multirow{2}{*}{
foobar
} & 
there
\\
\cline{2-2} & 
xxx.yyy.zzz.ttt
\\
\hline\end{tabulary}
```

I have a patch which creates rather this source:

``` latex
\noindent\begin{tabulary}{\linewidth}{|L|L|}
\hline
\multirow{2}{*}{\relax \textsf{\relax 
N
\unskip}\relax }\relax &\textsf{\relax 
Location Location
}\relax \\
\cline{2-2}&\textsf{\relax 
IP Address
}\relax \\
\hline\multirow{2}{*}{\relax 
foobar
\unskip}\relax &
there
\\
\cline{2-2}&
xxx.yyy.zzz.ttt
\\
\hline\end{tabulary}
```

**Comment by jfbu** [comment]
I used `\unskip` because I don't know where the `\n` is inserted after the multi-row content (here after foobar). It it was possible to prevent this and put the closing brace immediately after cell content that could possibly be better than doing `\unskip` on LaTeX side. But it must be sure there is no extra space there.

All these problems come from the fact that LaTeX books should devote a full chapter on how spaces in the source are treated, but they don't do it, and as a result anyone with some background in any other programming context has a wrong apprehension of the underlying issues.

The patch is at https://github.com/jfbu/sphinx/tree/fixmultirow, I did not know if the PR should go on stable or master (and I won't be able to follow the issue, perhaps, until some time). The topic branch is on top of current stable head.

Here is the diff. It is true that spaces are mostly ignored at begin and start of latex table cells, but if one starts including contents with spces in braces this becomes wrong. Spaces in TeX are not invisible as they also play both a crucial typographical rôle and a syntactic rôle on tokenization.

```
diff --git a/sphinx/writers/latex.py b/sphinx/writers/latex.py
index a3477c9..b57dac8 100644
--- a/sphinx/writers/latex.py
+++ b/sphinx/writers/latex.py
@@ -1123,26 +1123,26 @@ class LaTeXTranslator(nodes.NodeVisitor):
                     extracols = self.remember_multirowcol[self.table.col]
                     self.body.append(' \\multicolumn{')
                     self.body.append(str(extracols + 1))
-                    self.body.append('}{|l|}{}')
+                    self.body.append('}{|l|}{}\\relax ')
                     self.table.col += extracols
-                self.body.append(' & ')
+                self.body.append('&')
         else:
-            self.body.append(' & ')
+            self.body.append('&')
         self.table.col += 1
         context = ''
         if 'morecols' in node:
-            self.body.append(' \\multicolumn{')
+            self.body.append('\\multicolumn{')
             self.body.append(str(node.get('morecols') + 1))
             if self.table.col == 1:
-                self.body.append('}{|l|}{')
+                self.body.append('}{|l|}{\\relax ')
             else:
-                self.body.append('}{l|}{')
-            context += '}'
+                self.body.append('}{l|}{\\relax ')
+            context += '}\\relax '
         if 'morerows' in node:
-            self.body.append(' \\multirow{')
+            self.body.append('\\multirow{')
             self.body.append(str(node.get('morerows') + 1))
-            self.body.append('}{*}{')
-            context += '}'
+            self.body.append('}{*}{\\relax ')
+            context += '\\unskip}\\relax '
             self.remember_multirow[self.table.col] = node.get('morerows')
         if 'morecols' in node:
             if 'morerows' in node:
@@ -1154,22 +1154,22 @@ class LaTeXTranslator(nodes.NodeVisitor):
             self.literal_whitespace += 1
             self.body.append('\\eqparbox{%d}{\\vspace{.5\\baselineskip}\n' % id(node))
             self.pushbody([])
-            context += '}'
+            context += '}\\relax '
         if isinstance(node.parent.parent, nodes.thead):
             if len(node) == 1 and isinstance(node[0], nodes.paragraph) and node.astext() == '':
                 pass
             else:
                 self.body.append('\\textsf{\\relax ')
-                context += '}'
+                context += '}\\relax '
         while self.remember_multirow.get(self.table.col + 1, 0):
             self.table.col += 1
             self.remember_multirow[self.table.col] -= 1
-            context += ' & '
+            context += '&'
             if self.remember_multirowcol.get(self.table.col, 0):
                 extracols = self.remember_multirowcol[self.table.col]
-                context += ' \\multicolumn{'
+                context += '\\multicolumn{'
                 context += str(extracols + 1)
-                context += '}{l|}{}'
+                context += '}{l|}{}\\relax '
                 self.table.col += extracols
         if len(node.traverse(nodes.paragraph)) >= 2:
             self.table.has_problematic = True
```

**Comment by tk0miya** [comment]
> I used \unskip because I don't know where the \n is inserted after the multi-row content (here after foobar).

It is inserted at `visit_paragraph`. Your example are converted to following doctree (an internal data structure of sphinx). You can see the paragraph node under each entry nodes.

```
        <table>
            <tgroup cols="2">
                <colspec colwidth="8">
                <colspec colwidth="19">
                <thead>
                    <row>
                        <entry morerows="1">
                            <paragraph>
                                N
                        <entry>
                            <paragraph>
                                Location Location
                    <row>
                        <entry>
                            <paragraph>
                                IP Address
                <tbody>
                    <row>
                        <entry morerows="1">
                            <paragraph>
                                foobar
                        <entry>
                            <paragraph>
                                there
                    ...
```

(Note: sphinx does not recognize contents of cells as paragraph always.)

> I did not know if the PR should go on stable or master (and I won't be able to follow the issue, perhaps, until some time).

I usually consider on following things on commiting.
1) the change is a bug fix or not
2) the change is stable or not. Especially, latex writer is too fragile because users can combine nodes freely in reST side. Sometime it generates unexpected latex codes...

I think this is a bug fix. And I believe `\relax` and `\unskip` don't break the documents.
So I agree to go this into stable.

**Comment by jfbu** [comment]
thanks @tk0miya  for reviewing. I will look at it again in a few hours. There is one situation where `\unskip` can break a document, it is when it is issued at top level outside of a paragraph. But this can't be the case in a tabular cell, even if contents are in a minipage (inside a minipage TeX is in restricted vertical mode and one can do `\unskip` even outside of paragraphs; one can't do `\unskip` when TeX is in (global) vertical mode)

**Comment by jfbu** [comment]
Adding  a `\\relax` before the `\n` in 

``` python
    def depart_paragraph(self, node):
        self.body.append('\n')
```

could help prevent spurious space tokens in latex. But I am afraid of such drastic change. Hence I will stick with \unskip approach. 

Also we could change `visit_paragraph` but this is less needed from the extra `\relax` already added at https://github.com/jfbu/sphinx/tree/fixmultirow. I noticed on the diff above https://github.com/sphinx-doc/sphinx/issues/2723#issuecomment-229342954 that I had left a space before a `\multicolumn`. I will remove that although it does not matter I think.

**Comment by jfbu** [comment]
fixed at b25dc5e, thanks @tk0miya  for reviewing !
